test/HealthChecks.Elasticsearch.Tests/HealthChecks.Elasticsearch.approved.txt

```diff
@@ -32,5 +37,6 @@ namespace Microsoft.Extensions.DependencyInjection
     {
         public static Microsoft.Extensions.DependencyInjection.IHealthChecksBuilder AddElasticsearch(this Microsoft.Extensions.DependencyInjection.IHealthChecksBuilder builder, System.Action<HealthChecks.Elasticsearch.ElasticsearchOptions>? setup, string? name = null, Microsoft.Extensions.Diagnostics.HealthChecks.HealthStatus? failureStatus = default, System.Collections.Generic.IEnumerable<string>? tags = null, System.TimeSpan? timeout = default) { }
         public static Microsoft.Extensions.DependencyInjection.IHealthChecksBuilder AddElasticsearch(this Microsoft.Extensions.DependencyInjection.IHealthChecksBuilder builder, string elasticsearchUri, string? name = null, Microsoft.Extensions.Diagnostics.HealthChecks.HealthStatus? failureStatus = default, System.Collections.Generic.IEnumerable<string>? tags = null, System.TimeSpan? timeout = default) { }
+        public static Microsoft.Extensions.DependencyInjection.IHealthChecksBuilder AddElasticsearch(this Microsoft.Extensions.DependencyInjection.IHealthChecksBuilder builder, System.Func<System.IServiceProvider, Elastic.Clients.Elasticsearch.ElasticsearchClient>? clientFactory = null, System.Action<HealthChecks.Elasticsearch.ElasticsearchOptions>? setup = null, string? name = null, Microsoft.Extensions.Diagnostics.HealthChecks.HealthStatus? failureStatus = default, System.Collections.Generic.IEnumerable<string>? tags = null, System.TimeSpan? timeout = default) { }
```

Comment:
I’ve been trying to wrap my head around the shape of public API and got to the following conclusions.
The `ElasticsearchOptions` option bag is mutable and can be modified by the optional setup action (`System.Action<HealthChecks.Elasticsearch.ElasticsearchOptions>? setup = null`).
The setup action takes no parameter that would allow the user to read the information from the config (which is how I imagine users provide address, username, key or password).
```cs
services.AddHealthChecks().AddElasticsearch(setup =>
{
    // how should the user read these values here?
    setup.UseElasticCloud("cloudId", "cloudApiKey");
});
```
The `CheckHealthAsync` method performs a late validation of the option bag and throws if something is missing. 
https://github.com/Xabaril/AspNetCore.Diagnostics.HealthChecks/blob/4a49dfa83b0524661387c9ee7b9aeff120198c14/src/HealthChecks.Elasticsearch/ElasticsearchHealthCheck.cs#L33-L36
https://github.com/Xabaril/AspNetCore.Diagnostics.HealthChecks/blob/4a49dfa83b0524661387c9ee7b9aeff120198c14/src/HealthChecks.Elasticsearch/ElasticsearchOptions.cs#L34-L37
That exception won’t be easy to discover, as it will be reported somewhere up in the ASP.NET stack as a health check error.
On the other hand, the `ElasticsearchClientSettings`provided by Elastic has dedicated ctors that expect everything to be provided up front.
https://github.com/Xabaril/AspNetCore.Diagnostics.HealthChecks/blob/7b374ae80a32234a48edb438a5d519f686ce8de1/src/HealthChecks.Elasticsearch/ElasticsearchHealthCheck.cs#L35-L37

I believe that this overload should not expose the setup action and just require the users to configure everything before calling the factory method in the DI container and then just resolve the right instance in the setup

Example based on [this doc](https://www.elastic.co/guide/en/elasticsearch/client/net-api/7.16/function-as-a-service-environments.html)

```cs
builder.Services.AddSingleton(s => new ElasticClient(
    Environment.GetEnvironmentVariable("ELASTIC_CLOUD_ID"),
    new ApiKeyAuthenticationCredentials(Environment.GetEnvironmentVariable("ELASTIC_API_KEY"))));

builder.Services..AddHealthChecks().AddElasticsearch();
```

```suggestion
        public static Microsoft.Extensions.DependencyInjection.IHealthChecksBuilder AddElasticsearch(this Microsoft.Extensions.DependencyInjection.IHealthChecksBuilder builder, System.Func<System.IServiceProvider, Elastic.Clients.Elasticsearch.ElasticsearchClient>? clientFactory = null, string? name = null, Microsoft.Extensions.Diagnostics.HealthChecks.HealthStatus? failureStatus = default, System.Collections.Generic.IEnumerable<string>? tags = null, System.TimeSpan? timeout = default) { }
```